src/main/java/de/hpi/datastreams/processors/ServerProcessor.java

```diff
@@ -41,28 +41,27 @@ public void init(ProcessorContext context) {
      */
     @Override
     public void process(Long partitionKey, GradientMessage message) {
-        System.out.println("ServerProcessor received: " + message.toString());
+        System.out.println(String.format(
+                "ServerProcessor (partition %d) received: %s",
+                Math.toIntExact(partitionKey), message.toString()));
 
+        // Initialize ML model when receiving initial message
         if (message.getVectorClock().equals(START_VC)) {
             this.initializeModel();
         }
 
-        try {
-            Thread.sleep(1000);
-        } catch (InterruptedException e) {
-            e.printStackTrace();
-        }
-
+        // Extract gradients from message and update locally stored weights
         IntStream.range(message.getKeyRangeStart(), message.getKeyRangeEnd()).forEach(key -> {
             Optional<Float> gradient = message.getValue(key);
             gradient.ifPresent(partialGradient -> updateWeight(key, partialGradient));
         });
 
-        // TODO: start training iteration - sequential consistency model
+        // Start new training iteration on workers by sending update of weights
+        // TODO: Currently only eventual consistency model supported
         KeyRange keyRange = this.getKeyRange();
         WeightsMessage weights = new WeightsMessage(message.getVectorClock() + 1, keyRange, this.getWeights(keyRange));
 
-        this.weightsMessageProducer.send(new ProducerRecord<>(WEIGHTS_TOPIC, 0L, weights));
+        this.weightsMessageProducer.send(new ProducerRecord<>(WEIGHTS_TOPIC, message.getPartitionKey(), weights));
```

Comment:
Is this the point where we rely on the fact that we have four servers, but we only want one in the future?

Reply:
I had to insert a `partitionKey` attribute within the `GradientMessage` so we can make sure that we send back the `WeightsMessage` to the worker we received the `GradientMessage` from.
Keep in mind here that `partitionKey` in `ServerProcessor.process` will always be `0` due to the fact that all workers send their gradients to partitionKey `0L`.
We might have to rethink that concept in the future due to heavy load on the one `ServerProcessor`.